File: Uber/12/12.cpp
#include<iostream>
#include<queue>
#include<vector>

class Node {
       public:
	int data;
	Node* left;
	Node* right;
	Node() {
		left = NULL;
		right = NULL;
	}
};

bool isBST(Node* root) {
	static Node* prev = NULL;
	if (root) {
		if (!isBST(root->left)) return false;
		if (prev != NULL && root->data <= prev->data) return false;
		prev = root;
		return isBST(root->right);
	}
	return true;
}

int main() {
	int m, n;
	std::cin >> m >> n;
	std::vector<Node*> tree;
	tree.reserve(m);
	for (int i = 0; i < m; i++) {
		tree.push_back(new Node());
	}
	for(int i = 0;i < m;i++) {
		std::cin >> tree[i] -> data;
	}
	while (m--) {
		int idx, l, r;
		std::cin >> idx >> l >> r;
		if (l >= 0) tree[idx]->left = tree[l];
		if (r >= 0) tree[idx]->right = tree[r];
	}
	if (isBST(tree[n]))
		std::cout << "YES\n";
	else
		std::cout << "NO\n";
	return 0;
}

File: Uber/12/12.py
class newNode: 
    def __init__(self, data): 
        self.data = data
        self.left = None
        self.right = None
    
def isBST(root):
	global prev
	prev = None
	return isBSTUtil(root)

def isBSTUtil(root):
	global prev
	if root is None:
		return True
	if isBSTUtil(root.left) is False:
		return False
	if prev is not None and prev.data > root.data:
		return False
	prev = root
	return isBSTUtil(root.right)

m, n = map(int, input().split())
tree = []
for x in range(m):
    tree.append(newNode(x))
nodeValue = list(map(int, input().split()))
for x in range(m):
    tree[x].data = nodeValue[x]
for x in range(m):
    idx, l, r = map(int, input().split())
    if l >= 0:
        tree[idx].left = tree[l]
    if r >= 0:
        tree[idx].right = tree[r]
if isBST(tree[n]):
    print("YES")
else:
    print("NO")
